source/extensions/filters/network/mysql_proxy/mysql_terminal_filter.cc

```diff
@@ -0,0 +1,522 @@
+      err = MessageHelper::defaultError(
+          "mysql proxy is failed to auth to upstream due to auth plugin is not support (now "
+          "support mysql_native_password, mysql_old_password)");
+      break;
+    }
+    auto auth_method =
+        auth_switch.isOldAuthSwitch() ? AuthMethod::OldPassword : AuthMethod::NativePassword;
+    seed = auth_switch.getAuthPluginData();
+
+    auto switch_resp =
+        MessageHelper::encodeClientLogin(auth_method, parent.upstream_username_,
+                                         parent.upstream_password_, parent.connect_db_, seed);
+    switch_resp.setSeq(login_resp.getSeq() + 1);
+    parent.sendRemote(switch_resp);
+    parent.stepRemoteSession(login_resp.getSeq() + 2, MySQLSession::State::AuthSwitchMore);
+    break;
+  }
+  case MYSQL_RESP_OK:
+    onAuthSucc();
+    break;
+  case MYSQL_RESP_ERR:
+    ENVOY_LOG(info, "mysql proxy failed of auth, info {}",
+              dynamic_cast<ErrMessage&>(login_resp).getErrorMessage());
+    err = MessageHelper::defaultError(
+        fmt::format("mysql proxy is failed to auth upstream database server due to {}",
+                    dynamic_cast<ErrMessage&>(login_resp).getErrorMessage()));
+    break;
+  case MYSQL_RESP_MORE:
+  default:
+    ENVOY_LOG(info, "unhandled message resp code {}", login_resp.getRespCode());
+    err = MessageHelper::defaultError(fmt::format("mysql proxy is failed to auth upstream database "
+                                                  "server due to unhandled message resp code {}",
+                                                  login_resp.getRespCode()));
+    break;
+  }
+  if (err.has_value()) {
+    // send error message to client when proxy auth error
+    parent.onFailure(err.value(), parent.downstream_event_handler_->pending_response->getSeq());
+  }
+}
+
+void MySQLTerminalFilter::UpstreamEventHandler::onClientSwitchResponse(ClientSwitchResponse&) {
+  ENVOY_LOG(error, "mysql filter: upstream onClientSwitchResponse impossible callback is called");
+}
+
+void MySQLTerminalFilter::UpstreamEventHandler::onMoreClientLoginResponse(
+    ClientLoginResponse& resp) {
+  parent.onMoreClientLoginResponse(resp);
+
+  absl::optional<ErrMessage> err;
+
+  switch (resp.getRespCode()) {
+  case MYSQL_RESP_MORE:
+  case MYSQL_RESP_OK:
+    onAuthSucc();
+    break;
+  case MYSQL_RESP_ERR:
+    ENVOY_LOG(info, "mysql proxy failed of auth, info {} ",
+              dynamic_cast<ErrMessage&>(resp).getErrorMessage());
+    err = MessageHelper::defaultError(
+        fmt::format("mysql proxy is failed to auth upstream database server due to {}",
+                    dynamic_cast<ErrMessage&>(resp).getErrorMessage()));
+    break;
+  case MYSQL_RESP_AUTH_SWITCH:
+  default:
+    ENVOY_LOG(info, "unhandled message resp code {}", resp.getRespCode());
+    err = MessageHelper::defaultError(fmt::format("mysql proxy is failed to auth upstream database "
```

Comment:
mysql authentication with upstream server failed due to unhandled message resp code {}